### ISSUE TITLE
problem about locating low frequency component

### ISSUE THREAD

**Issue by momentum-coding** [lingeringlight/ALOFT#4]
img_abs_[:, h_start:h_start + h_crop, w_start:, :]   
Here i think is extracting low frequency component from frequency map,
 but why not '' img_abs_[:, h_start:h_start + h_crop, :w_crop, :]   ''? since low frequency component locates at the centre

**Comment by lingeringlight** [comment]
Hi, thanks for your attention. Note that we here use [torch.fft.rfft2()](https://pytorch.org/docs/stable/generated/torch.fft.rfft2.html) to computes the 2-dimensional discrete Fourier transform of real input, which omits the negative frequencies in the last dimension.
In this case, with [torch.fft.fftshift()](https://pytorch.org/docs/stable/generated/torch.fft.fftshift.html), the low-frequency component is shifted to the “right” of the rffted frequency map, thus we use img_abs_[:, h_start:h_start + h_crop, w_start:, :] to extract the low-frequency component. 
Alternatively, we can also utilize torch.fft.fft2() and torch.fft.fftshift() functions which shift the low-frequency component to the center.

Good question! Let me know if you have any further problems.

**Comment by dslisleedh** [comment]
Hi @lingeringlight 

First of all, thanks for the excellent research. I do have a question, though, regarding the extraction of low-frequency components.

In Pytorch's FFTSHIFT implementation, there's no check to see whether the input frequencies have been FFTed or RFFTed; it merely moves the negative parts (values at the middle and after the middle index) to the front, rather than shifting the low-frequency components to the right.

The Pytorch documentation states the same:
```
 Reorders n-dimensional FFT data, as provided by fft, to have negative frequency terms first.
```

In your implementation, the w dimension is split into two parts even though there is no negative part. This occurs because of the use of FFTSHIFT with dim=(1,2). Here is the code and visualizations for this:
```python
import numpy as np

import torch
import torch.nn as nn

from PIL import Image
import matplotlib.pyplot as plt
import matplotlib.patches as patches

k = 7
gaussian_kernel = torch.zeros((k, k))
for i in range(k):
    for j in range(k):
        gaussian_kernel[i, j] = np.exp(-((i - k//2)**2 + (j - k//2)**2) / 2)
gaussian_kernel = gaussian_kernel / gaussian_kernel.sum()
gaussian_kernel = gaussian_kernel.unsqueeze(0).unsqueeze(-1)

noise = torch.randn((1, k, k, 1)) * 0.001
noisy_kernel = gaussian_kernel + noise  # B H W C like ALFOT implementation

rfft_randn = torch.abs(torch.fft.rfft2(noisy_kernel, dim=[1, 2], norm='ortho'))
rfft_randn_shift = torch.fft.fftshift(rfft_randn, dim=[1, 2])

w = k // 2 + 1
h_crop = int(k * np.sqrt(0.5))
w_crop = int(w * np.sqrt(0.5))

h_start = k // 2 - h_crop // 2
w_start = w - w_crop

fig, ax = plt.subplots(1, 2, figsize=(10, 5))
ax[0].pcolormesh(rfft_randn[0, :, :, 0].numpy())
ax[0].set_title('only rfft')
ax[1].pcolormesh(rfft_randn_shift[0, :, :, 0].numpy())
ax[1].set_title('rfft + fftshift(dim=[1, 2])')

rect = patches.Rectangle((w_start, h_start), w_crop, h_crop, linewidth=3, edgecolor='r', facecolor='none')

ax[1].add_patch(rect)
````
![image](https://github.com/lingeringlight/ALOFT/assets/48064931/14658cfa-85e7-42a0-9ccf-ed2293214694)


Even though the code might seem incorrect, it works as intended when the kernel size is small, and actually, your model performs well. However, when the kernel size becomes larger, it might cause serious problems. Here's a visualization when the spatial size is 64:

![image](https://github.com/lingeringlight/ALOFT/assets/48064931/6e7e22c8-6da8-4d5c-ac1d-845b44024158)

I think a solution for this is to apply FFTSHIFT only to the h dimension and then crop starting from the front. Here's the revised code:

```python
rfft_randn = torch.abs(torch.fft.rfft2(noisy_kernel, dim=[1, 2], norm='ortho'))
rfft_randn_shift = torch.fft.fftshift(rfft_randn, dim=[1])

w = k // 2 + 1
h_crop = int(k * np.sqrt(0.5))
w_crop = int(w * np.sqrt(0.5))

h_start = k // 2 - h_crop // 2

fig, ax = plt.subplots(1, 2, figsize=(10, 5))
ax[0].pcolormesh(rfft_randn[0, :, :, 0].numpy())
ax[0].set_title('only rfft')
ax[1].pcolormesh(rfft_randn_shift[0, :, :, 0].numpy())
ax[1].set_title('rfft + fftshift(dim=[1])')

rect = patches.Rectangle((0, h_start), w_crop, h_crop, linewidth=3, edgecolor='r', facecolor='none')

ax[1].add_patch(rect)
```
![image](https://github.com/lingeringlight/ALOFT/assets/48064931/33525983-33c7-4a1b-93dc-1fd56f2c1b0e)

Thank you for your time. Please let me know if there's anything wrong with this comment or if you need further clarification.

**Comment by lingeringlight** [comment]
Thanks for your attentive consideration. In the preceding version of our code, an approximative technique was employed to extract low-frequency spectrum information for computational efficiency, and your refined revision can empower the model to more effectively handle large mask sizes. We express our sincere gratitude and have incorporated your invaluable suggestion into our revised code.

**Comment by ZhouCX117** [comment]
@lingeringlight Hi, how can I obtain high-frequency and low-frequency images corresponding to eq5?
